File: ts/project/.sdk/tm/swift/Sources/ProjectNameSDK/core/Context.swift
// ProjectName SDK - operation context.
//
// ctxmap is a loose Swift dictionary that carries a mix of typed pipeline
// objects (client, utility, entity, spec, ...) and loose Value maps (data,
// match, options, ...), mirroring the C# donor's Dictionary<string,object?>.

import Foundation

// ctxProp reads a value from the loose ctx dictionary.
func ctxProp(_ m: [String: Any?]?, _ key: String) -> Any? {
  guard let m = m, let v = m[key] else { return nil }
  return v
}

public final class Context {
  public var id: String = ""
  public var out: [String: Any?] = [:]
  public var ctrl: Control = Control()
  public var meta: VMap = VMap()
  public var client: ProjectNameSDK?
  public var utility: Utility?
  public var op: Operation?
  public var point: VMap?
  public var config: VMap?
  public var entopts: VMap?
  public var options: VMap?
  public var opmap: [String: Operation] = [:]
  public var response: Response?
  public var result: Result?
  public var spec: Spec?
  public var data: VMap = VMap()
  public var reqdata: VMap = VMap()
  public var match: VMap = VMap()
  public var reqmatch: VMap = VMap()
  public var entity: Entity?
  public var shared: VMap?

  public init(_ ctxmap: [String: Any?]?, _ basectx: Context?) {
    id = "C" + String(Int.random(in: 10_000_000...99_999_999))

    // Client
    if let sdk = ctxProp(ctxmap, "client") as? ProjectNameSDK { client = sdk }
    if client == nil, let b = basectx { client = b.client }

    // Utility
    if let util = ctxProp(ctxmap, "utility") as? Utility { utility = util }
    if utility == nil, let b = basectx { utility = b.utility }

    // Ctrl
    ctrl = Control()
    let rawctrl = ctxProp(ctxmap, "ctrl")
    if let c = rawctrl as? Control {
      ctrl = c
    } else if let cm = rawctrl as? VMap {
      if let t = cm.entries["throw"]?.asBool { ctrl.throwErr = t }
      if let e = cm.entries["explain"]?.asMap { ctrl.explain = e }
      if let a = cm.entries["actor"]?.asString { ctrl.actor = a }
      if let p = cm.entries["paging"]?.asMap { ctrl.paging = p }
    } else if let b = basectx {
      ctrl = b.ctrl
    }

    // Meta
    meta = VMap()
    if let mm = ctxProp(ctxmap, "meta") as? VMap {
      meta = mm
    } else if let b = basectx {
      meta = b.meta
    }

    // Config
    if let cfg = ctxProp(ctxmap, "config") as? VMap { config = cfg }
    if config == nil, let b = basectx { config = b.config }

    // Entopts
    if let eo = ctxProp(ctxmap, "entopts") as? VMap { entopts = eo }
    if entopts == nil, let b = basectx { entopts = b.entopts }

    // Options
    if let om = ctxProp(ctxmap, "options") as? VMap { options = om }
    if options == nil, let b = basectx { options = b.options }

    // Entity
    if let ent = ctxProp(ctxmap, "entity") as? Entity { entity = ent }
    if entity == nil, let b = basectx { entity = b.entity }

    // Shared
    if let sh = ctxProp(ctxmap, "shared") as? VMap { shared = sh }
    if shared == nil, let b = basectx { shared = b.shared }

    // Opmap
    if let opm = ctxProp(ctxmap, "opmap") as? [String: Operation] {
      opmap = opm
    } else if let b = basectx {
      opmap = b.opmap
    }

    // Data maps
    data = (ctxProp(ctxmap, "data") as? VMap) ?? VMap()
    reqdata = (ctxProp(ctxmap, "reqdata") as? VMap) ?? VMap()
    match = (ctxProp(ctxmap, "match") as? VMap) ?? VMap()
    reqmatch = (ctxProp(ctxmap, "reqmatch") as? VMap) ?? VMap()

    // Point
    if let tm = ctxProp(ctxmap, "point") as? VMap { point = tm }
    if point == nil, let b = basectx { point = b.point }

    // Spec
    if let sp = ctxProp(ctxmap, "spec") as? Spec { spec = sp }
    if spec == nil, let b = basectx { spec = b.spec }

    // Result
    if let res = ctxProp(ctxmap, "result") as? Result { result = res }
    if result == nil, let b = basectx { result = b.result }

    // Response
    if let resp = ctxProp(ctxmap, "response") as? Response { response = resp }
    if response == nil, let b = basectx { response = b.response }

    // Resolve operation
    let opname = ctxProp(ctxmap, "opname") as? String ?? ""
    op = resolveOp(opname)
  }

  private func resolveOp(_ opname: String) -> Operation {
    // Cache key `<entity>:<opname>` so two entities with the same op get
    // distinct cached Operations.
    let entname = entity?.getName() ?? ""
    let cacheKey = entname + ":" + opname

    if let cached = opmap[cacheKey] { return cached }

    if opname == "" {
      return Operation(VMap())
    }

    let store: Value = config == nil ? .noval : .map(config!)
    let opcfg = getpath(store, jtp("entity", entname, "op", opname))

    let input = (opname == "update" || opname == "create") ? "data" : "match"

    var points: VList = VList()
    if let ocm = opcfg.asMap, let tl = ocm.entries["points"]?.asList {
      points = tl
    }

    let opm = VMap()
    opm.entries["entity"] = .string(entname)
    opm.entries["name"] = .string(opname)
    opm.entries["input"] = .string(input)
    opm.entries["points"] = .list(points)

    let op = Operation(opm)
    opmap[cacheKey] = op
    return op
  }

  public func makeError(_ code: String, _ msg: String) -> ProjectNameError {
    return ProjectNameError(code, msg, self)
  }
}

File: ts/project/.sdk/tm/kotlin/core/Context.kt
package KOTLINPACKAGE.core

import java.util.concurrent.ThreadLocalRandom

import KOTLINPACKAGE.utility.struct.Struct

/** Per-operation context threaded through the pipeline and feature hooks. */
@Suppress("UNCHECKED_CAST")
class Context(ctxmap: MutableMap<String, Any?>?, basectx: Context?) {

  var id: String
  var out: MutableMap<String, Any?> = linkedMapOf()
  var ctrl: Control = Control()
  var meta: MutableMap<String, Any?> = linkedMapOf()
  var client: SdkClient? = null
  var utility: Utility? = null
  lateinit var op: Operation
  var point: MutableMap<String, Any?>? = null
  var config: MutableMap<String, Any?>? = null
  var entopts: MutableMap<String, Any?>? = null
  var options: MutableMap<String, Any?>? = null
  var opmap: MutableMap<String, Operation>? = null
  var response: Response? = null
  var result: Result? = null
  var spec: Spec? = null
  var data: MutableMap<String, Any?> = linkedMapOf()
  var reqdata: MutableMap<String, Any?> = linkedMapOf()
  var match: MutableMap<String, Any?> = linkedMapOf()
  var reqmatch: MutableMap<String, Any?> = linkedMapOf()
  var entity: Entity? = null
  var shared: MutableMap<String, Any?>? = null

  init {
    this.id = "C" + (ThreadLocalRandom.current().nextInt(90000000) + 10000000)

    // Client
    val c = Helpers.getCtxProp(ctxmap, "client")
    if (c is SdkClient) {
      this.client = c
    }
    if (this.client == null && basectx != null) {
      this.client = basectx.client
    }

    // Utility
    val u = Helpers.getCtxProp(ctxmap, "utility")
    if (u is Utility) {
      this.utility = u
    }
    if (this.utility == null && basectx != null) {
      this.utility = basectx.utility
    }

    // Ctrl
    this.ctrl = Control()
    val cv = Helpers.getCtxProp(ctxmap, "ctrl")
    if (cv != null) {
      if (cv is MutableMap<*, *>) {
        val cm = cv as MutableMap<String, Any?>
        val t = cm["throw"]
        if (t is Boolean) {
          this.ctrl.throwing = t
        }
        val e = cm["explain"]
        if (e is MutableMap<*, *>) {
          this.ctrl.explain = e as MutableMap<String, Any?>
        }
        val a = cm["actor"]
        if (a is String) {
          this.ctrl.actor = a
        }
        val p = cm["paging"]
        if (p is MutableMap<*, *>) {
          this.ctrl.paging = p as MutableMap<String, Any?>
        }
      } else if (cv is Control) {
        this.ctrl = cv
      }
    } else if (basectx != null) {
      this.ctrl = basectx.ctrl
    }

    // Meta
    this.meta = linkedMapOf()
    val m = Helpers.getCtxProp(ctxmap, "meta")
    if (m is MutableMap<*, *>) {
      this.meta = m as MutableMap<String, Any?>
    } else if (basectx != null) {
      this.meta = basectx.meta
    }

    // Config
    val cfg = Helpers.getCtxProp(ctxmap, "config")
    if (cfg is MutableMap<*, *>) {
      this.config = cfg as MutableMap<String, Any?>
    }
    if (this.config == null && basectx != null) {
      this.config = basectx.config
    }

    // Entopts
    val eo = Helpers.getCtxProp(ctxmap, "entopts")
    if (eo is MutableMap<*, *>) {
      this.entopts = eo as MutableMap<String, Any?>
    }
    if (this.entopts == null && basectx != null) {
      this.entopts = basectx.entopts
    }

    // Options
    val o = Helpers.getCtxProp(ctxmap, "options")
    if (o is MutableMap<*, *>) {
      this.options = o as MutableMap<String, Any?>
    }
    if (this.options == null && basectx != null) {
      this.options = basectx.options
    }

    // Entity
    val e = Helpers.getCtxProp(ctxmap, "entity")
    if (e is Entity) {
      this.entity = e
    }
    if (this.entity == null && basectx != null) {
      this.entity = basectx.entity
    }

    // Shared
    val s = Helpers.getCtxProp(ctxmap, "shared")
    if (s is MutableMap<*, *>) {
      this.shared = s as MutableMap<String, Any?>
    }
    if (this.shared == null && basectx != null) {
      this.shared = basectx.shared
    }

    // Opmap
    val om = Helpers.getCtxProp(ctxmap, "opmap")
    if (om is MutableMap<*, *>) {
      this.opmap = om as MutableMap<String, Operation>
    }
    if (this.opmap == null && basectx != null) {
      this.opmap = basectx.opmap
    }
    if (this.opmap == null) {
      this.opmap = linkedMapOf()
    }

    // Data
    this.data = Helpers.toMapAny(Helpers.getCtxProp(ctxmap, "data")) ?: linkedMapOf()
    this.reqdata = Helpers.toMapAny(Helpers.getCtxProp(ctxmap, "reqdata")) ?: linkedMapOf()
    this.match = Helpers.toMapAny(Helpers.getCtxProp(ctxmap, "match")) ?: linkedMapOf()
    this.reqmatch = Helpers.toMapAny(Helpers.getCtxProp(ctxmap, "reqmatch")) ?: linkedMapOf()

    // Point
    val t = Helpers.getCtxProp(ctxmap, "point")
    if (t is MutableMap<*, *>) {
      this.point = t as MutableMap<String, Any?>
    }
    if (this.point == null && basectx != null) {
      this.point = basectx.point
    }

    // Spec
    val sp = Helpers.getCtxProp(ctxmap, "spec")
    if (sp is Spec) {
      this.spec = sp
    }
    if (this.spec == null && basectx != null) {
      this.spec = basectx.spec
    }

    // Result
    val r = Helpers.getCtxProp(ctxmap, "result")
    if (r is Result) {
      this.result = r
    }
    if (this.result == null && basectx != null) {
      this.result = basectx.result
    }

    // Response
    val resp = Helpers.getCtxProp(ctxmap, "response")
    if (resp is Response) {
      this.response = resp
    }
    if (this.response == null && basectx != null) {
      this.response = basectx.response
    }

    // Resolve operation
    val opname = Helpers.getCtxProp(ctxmap, "opname")
    this.op = resolveOp(if (opname is String) opname else "")
  }

  private fun resolveOp(opname: String): Operation {
    // Cache key is `<entity>:<opname>` so two entities with the same op
    // (e.g. both have a "list") get distinct cached Operations.
    var entname = ""
    if (this.entity != null) {
      entname = this.entity!!.name
    }
    val cacheKey = "$entname:$opname"

    val cached = this.opmap!![cacheKey]
    if (cached != null) {
      return cached
    }

    if ("" == opname) {
      return Operation(linkedMapOf())
    }

    val opcfg = Struct.getpath(this.config, listOf("entity", entname, "op", opname))

    var input = "match"
    if ("update" == opname || "create" == opname) {
      input = "data"
    }

    var points: MutableList<Any?>? = null
    if (opcfg is MutableMap<*, *>) {
      val t = Struct.getprop(opcfg, "points")
      if (t is MutableList<*>) {
        points = t as MutableList<Any?>
      }
    }
    if (points == null) {
      points = mutableListOf()
    }

    val opdef = linkedMapOf<String, Any?>()
    opdef["entity"] = entname
    opdef["name"] = opname
    opdef["input"] = input
    opdef["points"] = points

    val op = Operation(opdef)

    this.opmap!![cacheKey] = op
    return op
  }

  fun makeError(code: String, msg: String): SdkError {
    return SdkError(code, msg, this)
  }
}
